### DISCUSSION TITLE
LazyVim's Go extras does not specify DAP adapter?

### DISCUSSION THREAD

**Discussion by fredrikaverpil** [LazyVim/LazyVim#1107] [Q&A] [answered]
I've enabled the [lang.go extras](https://www.lazyvim.org/plugins/extras/lang.go) and I haven't really added anything else and I can successfully run tests. But I'm getting this error whenever I hit `<leader>td` to debug the test:

```
  Error  07:41:53 notify.error Neotest Adapter doesn't support chosen strategy.
   Warn  07:41:53 notify.warn Neotest neotest-go: ...edrik/.local/share/LazyVim/lazy/neotest/lua/nio/init.lua:105: The coroutine failed with this message: 
...yVim/lazy/neotest/lua/neotest/client/strategies/init.lua:55: attempt to index local 'instance' (a nil value)
stack traceback:
	...yVim/lazy/neotest/lua/neotest/client/strategies/init.lua: in function 'run'
	...share/LazyVim/lazy/neotest/lua/neotest/client/runner.lua:129: in function '_run_spec'
	...share/LazyVim/lazy/neotest/lua/neotest/client/runner.lua:89: in function <...share/LazyVim/lazy/neotest/lua/neotest/client/runner.lua:88>
```

My guess that the chosen strategy not supported is "dap", as this is the command I'm running:

```lua
require('neotest').run.run({strategy=dap})
```

It seems no go DAP is part of the extras setup. Am I expected to add something outside of the extras to make debugging work properly (and if so, any recommendations?) or am I doing something wrong?
[Here's](https://github.com/fredrikaverpil/dotfiles/blob/main/nvim-lazyvim/lua/plugins/debug.lua) my debug setup and [here's](https://github.com/fredrikaverpil/dotfiles/blob/main/nvim-lazyvim/lua/plugins/test.lua) my test setup.

Example test that gives me this error:

```go
// main_test.go

package main

import (
  "testing"
)

func TestStuff(t *testing.T) {
  x := 1

  if x != 1 {
    t.Errorf("Expected 1, got %v", x)
  }
}
```

**Comment by dpetka2001** [comment]
I think they might have forgotten to add `nvim-dap-go` as dependency to `nvim-dap` and call its config function to set it up. It seems to me they just install the go debugger in `mason`.

**Reply by fredrikaverpil** [reply]
I tried [adding this to my local config](https://github.com/fredrikaverpil/dotfiles/commit/4096a8a8d92cf0d3d58cf7fdbe4efad7c79140d9) but I'm still seeing the same error. Do you see what I'm doing wrong?

**Reply by jyuan0** [reply]
Think there should have been a adapter added mason-nvim-dap. No idea what's wrong though.

**Reply by folke** [reply]
you're probably overriding the mason `ensure_intalled` option

**Reply by dpetka2001** [reply]
Seems like `neotest-go` doesn't provide dap configuration. See https://github.com/nvim-neotest/neotest-go/pull/13 and https://github.com/nvim-neotest/neotest-go/issues/12.

**Reply by fredrikaverpil** [reply]
> you're probably overriding the mason ensure_intalled option

Hm, no I don't think so, as `delve` gets installed by Mason when I enable the LazyVim go extras. I believe the problem is there is no debug adapter registered in the go extras.


> Seems like neotest-go doesn't provide dap configuration

Ah, so in addition to the LazyVim extras not defining the debug adapter (e.g. `nvim-dap-go`), you also cannot invoke the debugger with `require("neotest").run.run({strategy = "dap"})` because the strategy is not implemented in `neotest-go`?

___

I worked around all this for now by adding the `nvim-dap-go` plugin along with a custom keymap, as seen in [this commit](https://github.com/fredrikaverpil/dotfiles/commit/080263a10309560588bb389a248422b2693da3fe). Now I can debug Go tests just fine.

Hmm, since the Go extras is a bit incomplete here, should we at least consider adding `nvim-dap-go` to the extras @folke ?
Is there maybe a way of overridding `<leader>td` when you have a go file opened, so it runs `require('dap-go').debug_test()`?

**Reply by fredrikaverpil** [reply, marked as answer]
I [opened a PR](https://github.com/LazyVim/LazyVim/pull/1115) where I add `nvim-dap-go` to the Go extras. But I'm unclear on what the best way forward would be to hook this up with the LazyVim "Debug Nearest Test" command (`<leader>td`).

**Reply by folke** [reply]
You could add a buffer-local mapping in an `on_attach` of golsp.

**Reply by dpetka2001** [reply]
Could we use the latest `key` implementation in `lspconfig.opts.servers["server"]`? Aren't these supposed to be enabled on `LspAttach` event when the servers get set up? I thought that was the reason for lsp-server specific keymaps.

**Reply by fredrikaverpil** [reply]
@folke ok, I hope I did the right thing to add a buffer-local mapping in my PR now 😅

@dpetka2001 I'm not sure what this means... If you like, please comment with a code suggestion in the PR?

**Reply by folke** [reply]
@dpetka2001 yes of course! That's exactly what can be used for this :)

@fredrikaverpil Check how it was done for rust

**Reply by fredrikaverpil** [reply]
@folke ok. I'm not sure if I did what you had in mind, but I ~~moved the `on_attach` into the gopls server settings~~ used the `keys` in the gopls server settings.
